stl/inc/xutility

```diff
@@ -4187,6 +4187,313 @@ struct default_sentinel_t {};
+        requires random_access_iterator<_Iter>
+    {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(-_Diff <= _Length, "counted_iterator decrement beyond end of range");
+        if constexpr (_Offset_verifiable_v<_Iter>) {
+            _Current._Verify_offset(-_Diff);
+        }
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        _Current -= _Diff;
+        _Length += _Diff;
+        return *this;
+    }
+    // clang-format on
+
+    // [counted.iter.cmp]
+    template <common_with<_Iter> _Other>
+    friend constexpr bool operator==(const counted_iterator& _Left, const counted_iterator<_Other>& _Right) {
+        _Verify_Range(_Left, _Right);
+        return _Left._Length == _Right._Length;
+    }
+
+    friend constexpr bool operator==(const counted_iterator& _Left, default_sentinel_t) noexcept // strengthened
+    {
+        return _Left._Length == 0;
+    }
+
+    template <common_with<_Iter> _Other>
+    friend constexpr strong_ordering operator<=>(
+        const counted_iterator& _Left, const counted_iterator<_Other>& _Right) {
+        _Verify_Range(_Left, _Right);
+        return _Right._Length <=> _Left._Length;
+    }
+
+    // [counted.iter.cust]
+    // clang-format off
+    friend constexpr iter_rvalue_reference_t<_Iter> iter_move(const counted_iterator& _It) noexcept(
+        noexcept(_RANGES iter_move(_It.current)))
+       requires input_iterator<_Iter>
+    {
+        return _RANGES iter_move(_It.current);
+    }
+    // clang-format on
+
+    template <indirectly_swappable<_Iter> _Other>
+    friend constexpr void iter_swap(const counted_iterator& _Left, const counted_iterator<_Other>& _Right) noexcept(
+        noexcept(_RANGES iter_swap(_Left._Current, _Right.base()))) {
+        _RANGES iter_swap(_Left._Current, _Right.base());
+    }
+
+    template <class _Other, enable_if_t<_Range_verifiable_v<_Iter, _Other>, int> = 0>
+    friend constexpr void _Verify_range(const counted_iterator& _First, const counted_iterator<_Other>& _Last) {
+        _STD next(_STD addressof(_First._Current), _First._Length)
+            == _STD next(_STD addressof(_Last.base()), _Last.count());
+    }
+
+    constexpr void _Verify_range(const counted_iterator& _Right) {
+        if constexpr (forward_iterator<_Iter>) {
+            _STD next(_Current, _Length) == _STD next(_Right._Current, _Right._Length);
+        }
+    }
+
+    using _Prevent_inheriting_unwrap = counted_iterator;
+
+    template <class _Other = _Iter, enable_if_t<_Offset_verifiable_v<_Other>, int> = 0>
+    constexpr void _Verify_offset(const iter_difference_t<_Iter> _Off) const {
+        _STL_VERIFY(_Off < _Length, "Offset larger than counted_iterator length");
+    }
+
+    template <class _Other = _Iter, enable_if_t<_Unwrappable_v<const _Other&>, int> = 0>
+    _NODISCARD constexpr counted_iterator<_Unwrapped_t<const _Other&>> _Unwrapped() const& {
+        return static_cast<counted_iterator<_Unwrapped_t<const _Other&>>>(_Current._Unwrapped());

```

Comment:
Shouldn't this be something like `return counted_iterator<const _Iter&>{_Current._Unwrapped(), _Length};`?

Reply:
I honestly copied it from `move_iterator`. That said the return value is `counted_iterator<_Unwrapped_t<const _Other&>>` so...

Reply:
Yeah, this is actually a remnant of the horrible ages of SFINAE. Fixed